### PR TITLE
Dynamic member access and indexers are not safe to return by reference.

### PR DESCRIPTION
**Pull request by VSadov**
Fixes:#16947

Dynamic member accesses and index expressions can be passed by reference. It is done via passing a ref to a temp holding the result of dynamic operation.

Dynamic operations are not safe to return by reference  though.

### CODE REVIEW

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler  - please take a look

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefReturnTests.cs [hunk 1]
```diff
@@ -2835,5 +2835,103 @@ static void Main()
...
+
+        foo(ref d);
+    }
+
+    public static ref dynamic foo(ref dynamic d)
+    {
+        return ref bar(ref d.Length);
+    }
+
+    public static ref dynamic bar(ref dynamic d)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefReturnTests.cs]: Consider renaming methods. Perhaps just `F` and `G`.

Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefReturnTests.cs]: Yes.  Please avoid use of foo and bar in the code base.  It's a policheck flag.

**Comment by cston** [comment]
Other than test method names, LGTM.

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - one more review?

**Pull request merged**